keps/sig-storage/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1023 @@
+   `spec.ordinals` and `spec.updateStrategy.rollingUpdate.partition` are also respected.
+   e.g.:
+   - If `spec.updateStrategy.type` is `RollingUpdate`,
+     update the PVCs in the order from the largest ordinal to the smallest.
+     Only proceed to the next ordinal when all the PVCs of the previous ordinal
+     are compatible with the new `volumeClaimTemplate`.
+   - If `spec.updateStrategy.type` is `OnDelete`,
+     Only update the PVC when the Pod is deleted.
+
+
+### What PVC is compatible
+
+A PVC is compatible with the template if:
+- All the immutable fields match exactly; and
+- `metadata.labels` and `metadata.annotations` of PVC is a superset of the template; and
+- `status.capacity.storage` of PVC is greater than or equal to
+  the `spec.resources.requests.storage` of the template; and
+- `status.currentVolumeAttributesClassName` of PVC is equal to
+  the `spec.volumeAttributesClassName` of the template.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1: Batch Expand Volumes
+
+We're running a CI/CD system and the end-to-end automation is desired.
+To expand the volumes managed by a StatefulSet,
+we can just use the same pipeline that we are already using to updating the Pod.
+All the test, review, approval, and rollback process can be reused.
+
+#### Story 2: Migrating Between Storage Providers
+
+We decide to switch from home-made local storage to the storage provided by a cloud provider.
+We can not afford any downtime, so we don't want to delete and recreate the StatefulSet.
+Our app can automatically rebuild the data in the new storage from other replicas.
+So we update the `volumeClaimTemplate` of the StatefulSet,
+delete the PVC and Pod of one replica, let the controller re-create them,
+then monitor the rebuild process.
+Once the rebuild completes successfully, we proceed to the next replica.
+
+#### Story 3: Migrating Between Different Implementations of the Same Storage Provider
+
+Our storage provider has a new version that provides new features, but can not be upgraded in-place.
+We can prepare some new PersistentVolumes using the new version, but referencing the same disk
+from the provider as the in-use PVs.
+Then the same update process as Story 2 can be used.
+Although the PVCs are recreated, the data is preserved, so no rebuild is needed.
+
+#### Story 4: Shinking the PV by Re-creating PVC
+
+After running our app for a while, we optimize the data layout and reduce the required storage size.
+Now we want to shrink the PVs to save cost.
+The same process as Story 2 can be used.
+
+#### Story 5: Asymmetric Replicas
+
+The replicas of our StatefulSet are not identical, so we still want to update
```

Comment:
```suggestion
The templates for different replicas of a StatefulSet are not identical, so we still want to update
```

Reply:
I would like the word "template" to specifically refer to the one written in StatefulSet. So I don't like "templates for different replicas". Maybe "The storage requirement of different replicas"?

Reply:
that's fine.